blog/_src/posts/2016-05-18-gradual-typing-across-the-spectrum.md

```diff
@@ -0,0 +1,181 @@
+(also see the upcoming [PLDI paper](http://www.cs.umd.edu/~jfoster/papers/pldi16.pdf) by
+Brianna and Jeff Foster).
+The basic idea is that it's hard to implement a traditional static typechecker for a language that
+heavily relies on metaprogramming, in which the fields/methods of classes
+may be rewritten at run-time. This is particularly tricky for frameworks like Ruby on Rails.
+Instead of checking types at compile-time, Hummingbird actually executes the typechecker
+at run-time in order to be able to accurately check programs that use run-time metaprogramming.
+To reduce overheads, she uses a cache for typechecking that is invalidated when classes are
+modified.
+
+Stephen Chang gave a very different view on metaprogramming in his talk, which focused on
+_implementing_ typecheckers using metaprogramming (the [trivial](http://docs.racket-lang.org/trivial/index.html)
+Typed Racket library is an offshoot of this work). His key idea is that typecheckers share
+many aspects with macro-based metaprogramming systems, such as the need to traverse syntax
+and annotate it with information. Since they share so much in common, why not just implement
+the typechecker as a macro? Stephen demonstrates that not only is this possible, but that
+it's possible to implement a wide variety of type system features this way including (local) type
+inference. The connection to gradual typing is that even a gradual type system can be
+implemented as a metaprogram by integrating the generation of dynamic checks into the macro
+transformation process.
+
+The last talk of the day (but certainly not the least), was by Michael Vitousek, who focused
+on the _transient_ implementation of gradual typing (first described in his
+[DLS paper](http://homes.soic.indiana.edu/mvitouse/papers/dls14.pdf)). Traditionally, gradual type systems
+have implemented their dynamic checks using [proxy](https://en.wikipedia.org/wiki/Proxy_pattern)
+objects that wrap method implementations with both pre- and post-checks. Unfortunately, this
+implementation technique often conflicts with the underlying language. Since proxying changes
+the identity of an object, this can interfere with object equality tests. Instead, the transient
+approach bakes the dynamic checks into and throughout the typed code to implement a "defense in depth"
+against inconsistencies with untyped code. The great thing about this implementation
+technique is that it doesn't demand any specialized support from the underlying language runtime
+and is therefore easy to port to other languages (like JavaScript).
+
+---
+
+That concludes the summary. Hopefully this helps provide a better picture of the state
+of gradual typing research. The exciting thing about gradual typing is that it contains
+both interesting theoretical problems and also connects to the practical needs of
+software developers. I'm hopeful that the Gradual Typing Across the Spectrum effort
+will have an impact on both academia and the software industry.
```

Comment:
The last phrase is a bit meh. I think the conclusion would be just as good without it.